awscli/customizations/s3/s3handler.py

```diff
@@ -513,3 +521,197 @@ def _process_transfer(self, future):
+
+    def _do_submit(self, fileinfo):
+        extra_args = {}
+        self.REQUEST_MAPPER_METHOD(extra_args, self._cli_params)
+        subscribers = []
+        self._add_additional_subscribers(subscribers, fileinfo)
+        # The result subscriber class should always be the last registered
+        # subscriber to ensure it is not missing any information that
+        # may have been added in a different subscriber such as size.
+        subscribers.append(self.RESULT_SUBSCRIBER_CLASS(self._result_queue))
+        self._submit_transfer_request(fileinfo, extra_args, subscribers)
+
+    def _add_additional_subscribers(self, subscribers, fileinfo):
+        pass
+
+    def _submit_transfer_request(self, fileinfo, extra_args, subscribers):
+        raise NotImplementedError('_submit_transfer_request()')
+
+    def _warn_and_signal_if_skip(self, fileinfo):
+        for warning_handler in self._get_warning_handlers():
+            if warning_handler(fileinfo):
+                # On the first warning handler that returns a signal to skip
+                # immediately propogate this signal and no longer check
+                # the other warning handlers as no matter what the file will
+                # be skipped.
+                return True
+
+    def _get_warning_handlers(self):
+        # Returns a list of warning handlers, which are callables that
+        # take in a single parameter representing a FileInfo. It will then
+        # add a warning to result_queue if needed and return True if
+        # that FileInfo should be skipped.
+        return []
+
+    def _should_inject_content_type(self):
+        return (
+            self._cli_params.get('guess_mime_type') and
+            not self._cli_params.get('content_type')
+        )
+
+    def _warn_glacier(self, fileinfo):
+        if not self._cli_params.get('force_glacier_transfer'):
+            if not fileinfo.is_glacier_compatible():
+                LOGGER.debug(
+                    'Encountered glacier object s3://%s. Not performing '
+                    '%s on object.' % (fileinfo.src, fileinfo.operation_name))
+                if not self._cli_params.get('ignore_glacier_warnings'):
+                    warning = create_warning(
+                        's3://'+fileinfo.src,
+                        'Object is of storage class GLACIER. Unable to '
+                        'perform %s operations on GLACIER objects. You must '
+                        'restore the object to be able to the perform '
+                        'operation.' %
+                        fileinfo.operation_name
+                    )
+                    self._result_queue.put(warning)
+                return True
+        return False
+
+
+class UploadRequestSubmitter(BaseTransferRequestSubmitter):
+    REQUEST_MAPPER_METHOD = RequestParamsMapper.map_put_object_params
+    RESULT_SUBSCRIBER_CLASS = UploadResultSubscriber
+
+    def is_valid_submission(self, fileinfo):
```

Comment:
Yeah that could work for streams, but it still would not work for moves. I am not sure if I will be able to ever rely on an operation name to transfer request submitter. 

And yes with the design right now, I just need to make sure the streams are the highest up on the resolution chain.

In general if there was one big resolver chain I could see the following transfer request submitters:

* upload
* upload stream
* move upload
* download
* download stream
* move download
* copy
* copy download
* delete

So that is nine in all. Not sure if this will grow with like the addition of dryruns but I can easily seen nine at the least.